subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/codegen/ApiTypeProvider.kt

```diff
@@ -0,0 +1,528 @@
+
+    override fun visitInnerClassType(localName: String) {
+        binaryName += "${'$'}$localName"
+    }
+
+    override fun visitTypeArgument() {
+        typeArguments.add(TypeSignatureVisitor().also { it.binaryName = "?" })
+    }
+
+    override fun visitTypeArgument(wildcard: Char): SignatureVisitor =
+        TypeSignatureVisitor().also {
+            expectTypeArgument = true
+            typeArguments.add(it)
+        }
+
+    override fun visitTypeVariable(internalName: String) {
+        visitBinaryName(binaryNameOfInternalName(internalName))
+    }
+
+    private
+    fun visitBinaryName(binaryName: String) {
+        if (expectTypeArgument) {
+            TypeSignatureVisitor().let {
+                typeArguments.add(it)
+                SignatureReader(binaryName).accept(it)
+            }
+            expectTypeArgument = false
+        } else {
+            this.binaryName = binaryName
+        }
+    }
+}
+
+
+private
+fun binaryNameOfBaseType(descriptor: Char) =
+    Type.getType(descriptor.toString()).className
+
+
+private
+fun binaryNameOfInternalName(internalName: String): String =
+    Type.getObjectType(internalName).className
+
+
+private
+fun sourceNameOfBinaryName(binaryName: String): String =
+    when (binaryName) {
+        "void" -> "Unit"
+        "?" -> "*"
+        in mappedTypeStrings.keys -> mappedTypeStrings[binaryName]!!
+        in primitiveTypeStrings.keys -> primitiveTypeStrings[binaryName]!!
+        else -> binaryName.replace('$', '.')
+    }
+
+
+/**
+ * See https://kotlinlang.org/docs/reference/java-interop.html#mapped-types
+ */
+private
+val mappedTypeStrings =
+    mapOf(
+        // Built-ins
+        "java.lang.Cloneable" to "kotlin.Cloneable",
+        "java.lang.Comparable" to "kotlin.Comparable",
+        "java.lang.Enum" to "kotlin.Enum",
+        "java.lang.Annotation" to "kotlin.Annotation",
+        "java.lang.Deprecated" to "kotlin.Deprecated",
+        "java.lang.CharSequence" to "kotlin.CharSequence",
+        "java.lang.Number" to "kotlin.Number",
+        "java.lang.Throwable" to "kotlin.Throwable",
+        // Collections
+        "java.util.Iterable" to "kotlin.collections.Iterable",
+        "java.util.Iterator" to "kotlin.collections.Iterator",
+        "java.util.ListIterator" to "kotlin.collections.ListIterator",
+        "java.util.Collection" to "kotlin.collections.Collection",
+        "java.util.List" to "kotlin.collections.List",
+        "java.util.ArrayList" to "kotlin.collections.ArrayList",
+        "java.util.Set" to "kotlin.collections.Set",
+        "java.util.HashSet" to "kotlin.collections.HashSet",
+        "java.util.LinkedHashSet" to "kotlin.collections.LinkedHashSet",
+        "java.util.Map" to "kotlin.collections.Map",
+        "java.util.Map.Entry" to "kotlin.collections.Map.Entry",
+        "java.util.HashMap" to "kotlin.collections.HashMap",
+        "java.util.LinkedHashMap" to "kotlin.collections.LinkedHashMap")
```

Comment:
:o: I would swear that this same map already exists somewhere in the codebase somewhere. I remember reviewing a PR with it. Should this really be duplicated here?

Reply:
You did see it in #815 that was closed and not merged.

Reply:
Okay, thanks.